controllers/azurejson_machine_controller.go

```diff
@@ -0,0 +1,201 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
+)
+
+// AzureJSONMachineReconciler reconciles azure json secrets for AzureMachine objects
+type AzureJSONMachineReconciler struct {
+	client.Client
+	Log              logr.Logger
+	Recorder         record.EventRecorder
+	ReconcileTimeout time.Duration
+}
+
+// SetupWithManager initializes this controller with a manager
+func (r *AzureJSONMachineReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.AzureMachine{}).
+		WithEventFilter(filterUnclonedMachinesPredicate{log: r.Log}).
+		Owns(&corev1.Secret{}).
+		Complete(r)
+}
+
+type filterUnclonedMachinesPredicate struct {
+	log logr.Logger
+	predicate.Funcs
+}
+
+func (f filterUnclonedMachinesPredicate) Create(e event.CreateEvent) bool {
+	return f.Generic(event.GenericEvent{
+		Meta:   e.Meta,
+		Object: e.Object,
+	})
+}
+
+func (f filterUnclonedMachinesPredicate) Update(e event.UpdateEvent) bool {
+	return f.Generic(event.GenericEvent{
+		Meta:   e.MetaNew,
+		Object: e.ObjectNew,
+	})
+}
+
+// Generic implements default GenericEvent filter
+func (f filterUnclonedMachinesPredicate) Generic(e event.GenericEvent) bool {
+	if e.Object == nil {
+		f.log.Error(nil, "Generic event has no old metadata", "event", e)
+		return false
+	}
+
+	if e.Meta == nil {
+		f.log.Error(nil, "Generic event has no new metadata", "event", e)
+		return false
+	}
+
+	_, isControlPlaneNode := e.Meta.GetLabels()[clusterv1.MachineControlPlaneLabelName]
+	_, isMachineSetNode := e.Meta.GetLabels()[clusterv1.MachineSetLabelName]
+	_, isMachineDeploymentNode := e.Meta.GetLabels()[clusterv1.MachineDeploymentLabelName]
+
+	return !isControlPlaneNode && !isMachineSetNode && !isMachineDeploymentNode
+}
+
+// Reconcile reconciles the azure json for a specific machine not in a machine deployment
```

Comment:
why does this need to be distinct from machine deployments? I'm sure there's a good reason but I don't think I'm completely following the logic

Reply:
any machines created from KCP, machinset or machinedeployment will already have a secret created for them from the template controller. Those machines share 1 secret per template (all machines in the MD reuse the same secret). Machinepools don't use templates; they need a separate, unfiltered controller. 

if a user creates a machine one-off, without machineset or machinedeployment, we need to provision them an azure-json. This controller catches that case, while ignoring other machines.

Reply:
can add a comment to that effect